codegen/smithy-ruby-rails-codegen/middleware/request_id.rb

```diff
@@ -0,0 +1,16 @@
+module Middleware
+  # A middleware that extracts a request id from a response and sets it
+  # on the context.
+  # @api private
+  class RequestId
+    def initialize(app)
+      @app = app
+    end
+
+    def call(input, context)
+      output = @app.call(input, context)
+      output.metadata[:request_id] = context.response.headers['x-request-id']
```

Comment:
I don't like that I had to parse here (for data) and in error parser (for errors). I'd like one place to put this information.

Reply:
I agree - I'd rather be able to do this in a single place.  But I don't believe its possible without some interface tweak, either adding output to the error parser interface (to allow access to metadata) or to add write access to the request_id.

We can add either of those later if we want (but taking them away would be hard).  So, lets leave this for now and accept some duplication.

Reply:
Talked offline, metadata is now a property of errors.